### PR TITLE
Support quadratic constraints

### LINKED ISSUE #50: Add support for quadratic constraints
It could be useful to treat separately quadratic and other nonlinear constraints.
An open question is how to store all hessian of the quadratic constraints?
- List of 2D COO
- 3D COO for third order tensor (rows, cols, tubes, vals)
- ...

### LINKED ISSUE #149: Improve hess_coord! when y is not provided
https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/pull/130

### LINKED ISSUE #150: Unsupported features should error, not warn
@ccoffrin was running a few examples and getting incorrect results because quadratic constraints are not supported (#50).

Shouldn't this be an error instead of a warning?

https://github.com/JuliaSmoothOptimizers/NLPModelsJuMP.jl/blob/12d8f2315328f1b8c8af0a6a4552b6ad71327674/src/utils.jl#L216-L219

### PR DESCRIPTION
**Pull request by amontoison**
close #50
close #149 
close #150
@tmigot You should be happy now :smiley:

### CODE REVIEW

**Comment by amontoison** [comment]
@tmigot We should probably add HS61, HS100 and HS219 in `NLPModelsTest.jl` for more robust tests.

**Comment by amontoison** [comment]
If we replace `@NLconstraint` with `@constraint` for the quadratic constraints in `hs6`, `hs10`, `hs11`, and `hs14`, almost all tests pass.
The failing tests are related to this issue: [JuliaSmoothOptimizers/NLPModelsTest.jl#109](https://github.com/JuliaSmoothOptimizers/NLPModelsTest.jl/issues/109).

**Review by tmigot — CHANGES_REQUESTED** [review]
Hi @amontoison ! Thanks for the PR that's great.
I just had minor comments to add some tests.
How did you verify the result?

File: src/utils.jl [hunk 1]
```diff
@@ -419,8 +586,8 @@ function parser_objective_MOI(moimodel, nvar, index_map)
         push!(rows, i)
         push!(cols, j)
       else
-        push!(cols, j)
-        push!(rows, i)
+        push!(rows, j)
+        push!(cols, i)
       end
```
[the remark this answers is not in the document]
Comment (amontoison) [reply src/utils.jl:591]: Yes, I did a typo here a long time ago but by chance JuMP seems to only stored the upper triangular part of the symmetric matrix. We never went in the "else" block.

File: src/utils.jl [hunk 2]
```diff
@@ -237,13 +388,29 @@ function parser_MOI(moimodel, index_map)
...
+  quad_nnzj = 0
+  quad_nnzh = 0
+  for i = 1:nquad
+    quad_nnzj += qcons[i].nnzg
+    quad_nnzh += qcons[i].nnzh
+  end
+  quadcon = QuadraticConstraints(nquad, qcons, quad_nnzj, quad_nnzh)
 
-  return nlin, lincon, lin_lcon, lin_ucon
+  return nlin, lincon, lin_lcon, lin_ucon, quadcon, quad_lcon, quad_ucon
```
[the remark this answers is not in the document]
Comment (amontoison) [reply src/utils.jl:413]: We can't store `nquad` in the `NLPMeta` structure like `nlin` so I stored it in `quadcon`.

File: test/nlp_problems/hs100.jl [hunk 3]
```diff
@@ -0,0 +1,24 @@
+function hs100(args...; kwargs...)
+  nlp = Model()
+  x0 = [1, 2, 0, 4, 0, 1, 1]
+  @variable(nlp, x[i = 1:7], start = x0[i])
+
+  @NLconstraint(nlp, 127 - 2 * x[1]^2 - 3 * x[2]^4 - x[3] - 4 * x[4]^2 - 5 * x[5] ≥ 0)
+  @constraint(nlp, 282 - 7 * x[1] - 3 * x[2] - 10 * x[3]^2 - x[4] + x[5] ≥ 0)
+  @constraint(nlp, 196 - 23 * x[1] - x[2]^2 - 6 * x[6]^2 + 8 * x[7] ≥ 0)
+  @constraint(nlp, -4 * x[1]^2 - x[2]^2 + 3 * x[1] * x[2] - 2 * x[3]^2 - 5 * x[6] + 11 * x[7] ≥ 0)
```
[the remark this answers is not in the document]
Comment (amontoison) [reply test/nlp_problems/hs100.jl]: I replaced
```julia
@constraint(nlp, 196 - 23 * x[1] - x[2]^2 - 6 * x[6]^2 + 8 * x[7] ≥ 0)
```
by
```julia
@constraint(nlp, -196 + 23 * x[1] + x[2]^2 + 6 * x[6]^2 - 8 * x[7] ≤ 0)
```

**Review by tmigot — APPROVED** [review]
Thanks @amontoison ! It looks good to me.
I think we just need a problem in NLPModelsTest that has both quadratic and nonlinear constraints.
Let's go for 0.13.0 when you are ready.

File: src/utils.jl [hunk 4]
```diff
@@ -200,11 +224,132 @@ function parser_VAF(fun, set, linrows, lincols, linvals, nlin, lin_lcon, lin_uco
...
+  push!(qcons, qcon)
+end
+
+"""
+    parser_VQF(fun, set, nvar, qcons, quad_lcon, quad_ucon, index_map)
+
+Parse a `VectorQuadraticFunction` fun with its associated set.
+`qcons`, `quad_lcon`, `quad_ucon` are updated.
+"""
+function parser_VQF(fun, set, nvar, qcons, quad_lcon, quad_ucon, index_map)
```
[the remark this answers is not in the document]
Comment (amontoison) [reply src/utils.jl:290]: Done!

**Pull request merged**